SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp

```diff
@@ -11,31 +13,64 @@ DEFINE_LOG_CATEGORY(LogSpatialVirtualWorkerTranslator);
 SpatialVirtualWorkerTranslator::SpatialVirtualWorkerTranslator()
 	: NetDriver(nullptr)
 {
+	bWorkerEntityQueryInFlight = false;
+	bIsReady = false;
+	DesiredVirtualWorkerCount = 0;
 }
 
 void SpatialVirtualWorkerTranslator::Init(USpatialNetDriver* InNetDriver)
 {
 	NetDriver = InNetDriver;
+	if (NetDriver != nullptr)
+	{
+		WorkerId = NetDriver->Connection->GetWorkerId();
+	}
+	else
+	{
+		WorkerId = "InvalidWorkerId";
+	}
 }
 
-const FString* SpatialVirtualWorkerTranslator::GetPhysicalWorkerForVirtualWorker(VirtualWorkerId id)
+void SpatialVirtualWorkerTranslator::SetDesiredVirtualWorkerCount(uint32 NumberOfVirtualWorkers)
 {
-	return VirtualToPhysicalWorkerMapping.Find(id);
-}
+	// Currently, this should only be called once on startup. In the future we may allow for more
+	// flexibility. 
+	if (bIsReady)
+	{
+		UE_LOG(LogSpatialVirtualWorkerTranslator, Warning, TEXT("(%s) SetDesiredVirtualWorkerCount called after the translator is ready, ignoring."), *WorkerId);
+	}
 
-void SpatialVirtualWorkerTranslator::ApplyVirtualWorkerManagerData(const Worker_ComponentData& Data)
-{
-	if (NetDriver)
+	DesiredVirtualWorkerCount = NumberOfVirtualWorkers;
+	for (uint32 i = 0; i < NumberOfVirtualWorkers; i++)
 	{
-		UE_LOG(LogSpatialVirtualWorkerTranslator, Log, TEXT("(%s) ApplyVirtualWorkerManagerData"), *NetDriver->Connection->GetWorkerId());
+		UnassignedVirtualWorkers.Enqueue(i);
 	}
 
-	Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data.schema_type);
+	// TODO(zoning): We likely want to not declare ready until we have received the connection entity query and have
+	// synchronized with the strategy and enforcer.
+	bIsReady = true;
+}
 
-	VirtualToPhysicalWorkerMapping.Empty();
+
+const FString* SpatialVirtualWorkerTranslator::GetPhysicalWorkerForVirtualWorker(VirtualWorkerId id)
+{
+	return VirtualToPhysicalWorkerMapping.Find(id);
+}
+
+void SpatialVirtualWorkerTranslator::ApplyVirtualWorkerManagerData(Schema_Object* ComponentObject)
```

Comment:
Do we modify ComponentObject? Otherwise perhaps const it?

Reply:
Actually, turns out it can't be const because the Worker SDK code needs a non-const.